README
Class-AutoDB

This class works closely with Class::AutoClass to provide almost
transparent object persistence.  You can declare any class to be
persistent and can easily store any objects of a persistent class.
You can retrieve a stored object explicitly through a query
mechansism, or implicitly by following a reference pointing to the
object.  The persistence mechanism preserves the structure of the
object network: if objects A and B pointed to C when they were stored,
the retrieval process will reconstruct this relationship.

This class presently works only with MySQL.  

CAUTION: This module is old. We use it internally, and while it works
well for our purposes, we urge new users to heed the warnings in the
documentation.  This release brings the CPAN version of the module
up-to-date relative to our internal version, something we should have
done long ago.  

We are resuming active development of the code and expect to release
improved versions over the next several months.  See the "Roadmap"
section in the main documentation.

The 't' directory includes separate 'user' and 'developer' test
suites.  The installation procedure below runs the user test suite.
See t/README for more information.

The 'docs' directory contains additional documentation.  All of this
is presently for internal classes.

INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

The module includes a modified version of Data::Dumper, renamed
Class::AutoDB::Dumper.  We include both the pure Perl and much faster
C (.xs) implementations. You need a C compiler to compile the C
implementation. We get a few compiler warnings when compiling
Dumper.xs. These seem benign and can be ignored. The Data::Dumper docs
claim that the code will fall back to the pure Perl implementation if
the C version is not available.  We haven't checked this claim.

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Class::AutoDB

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Class-AutoDB

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Class-AutoDB

    CPAN Ratings
        http://cpanratings.perl.org/d/Class-AutoDB

    Search CPAN
        http://search.cpan.org/dist/Class-AutoDB/


COPYRIGHT AND LICENCE

Copyright (C) 2009 Nat Goodman

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Class-AutoDB

ACTIVE MAINTENANCE and DEVELOPMENT ENDED!
       While debugging 1.28 on Cantrell's machine, the compile of our patched
       version of Data::Dumper generated 'deprecated' messages.
       To continue work on this package, we will need to redo the Dumper patch.
       Too much work for too little gain.

1.29	2013-12-18
	Changes to tests only
        FAILs seen by David Cantrell on 1.28 caused by running of multiple test
	instances in parallel. All tests were using same database and trampling
	on each other. Solution is to create unique database for each test run. 
	There may also be cases where tester cannot create new databases. For
	such cases, tests use preexisting 'test' database.
	BUT...
	While developing these changes, the only two CPAN testers that were
	previously able to run the test suite (Slaven Rezić and David Cantrell)
	no longer can do so.  
	- For Cantrell, the issue I think is that the new version of DBD::mysql
	  (4.025) is FAILing its tests, and Cantrell's tester also tries to
	  install the latest version of dependencies.
	- For Rezić, I have no clue.

1.28	2013-10-28
	Fixed problems with manifest introduced in 1.27. Added dependency for
	Test::CheckManifest
	Unable to reproduce FAILs seen by David Cantrell.
	Changes to tests
	- streamlined manifest test
	- made the socket test more cautious when connecting to a private MySQL
	  instance. Tries to connect as user, then as root, then gives up 
	- some tests were connecting to MySQL using undef as the user instead of
	  $ENV{USER}. These are usually equivalent but may not be if the 
	  DBI_USER environment variable is set

1.27	2013-10-05
	Bug fixes
	- connect was using the wrong property name when setting 'socket' in
	  dsn. was using 'sock'; should be 'mysql_socket'
	Updated depdendencies
	Changes to tests
	 - regression test for socket bug
	 - updated 002.pod to strip development sub-version number
	 - imported 003.manifest from Babel
         - fixed boilerplate & pod tests
	 - added 000.reqs test to check existence of MySQL database, adapted 
	   from Babel
         - removed Developer tests from distro. Haven't been tested in a while!

1.26	2012-11-29
	Changes to tests
	- fixed the change made in 1.25.  $^V, the variable containing Perl's
	  version, works differently in different versions. in perls > 5.10 it
	  returns a regular string, while in earlier versions it returns a 
	  v-string. naturally, the code to handle these cases is different...

1.25	2012-11-28
	Changes to tests
	- tests of DOES method now check for Perl version >= 5.10.1, since that
	  is when DOES was added to UNIVERSAL. thanks to the CPAN smoke testers
	  for finding this problem! 

1.24	2012-10-28
	Added test for version number in POD
	Removed redundant 'dist_version_from' param from Build.PL
	Bug fixes
	- when invoking a UNIVERSAL method (isa, can, DOES, VERSION) on Oid or
	  OidDelelted, the method was not redispatched to the real class

1.23    2012-08-02
	Changed code and tests to correct deprecated use of qw(...) as parentheses
	Changes to Build.PL
	- added Module::Build to configure_requires
        - added dist_version_from

1.22	2011-01-07
	Bug fixes
	- when 'using' a class at runtime, schema changes were not happening;
	  the common failure case was a class loaded at runtime that defines
	  a collection
	Changes to Build.PL
	- added explicit test for DBD::mysql based on code in Data::Babel's
	  Build.PL and removed from prereqs. it doesn't work to put in prereqs
	  because if not present, install tries to install 'DBD' which does not 
	  exist
	Changes to tests
	- declare.030.pct_autodb.t was not setting 'create' flag in call to
	  'new'. longstanding bug that remained undetected because previous
	  tests left the database in a state that didn't bother it

1.21	2011-01-05
	Bug fixes;
	- when deleting objects whose classes were not yet 'used', the deletes 
	  were not being applied to the objects' collections; this was because
	  the class-to-collection information was set in the registry

1.20	2010-09-20
        New features
	- 'del' method to delete objects
        - queries ('get', 'find', 'count') with no arguments retrieve/count all
	  objects. makes it possible to retrieve objects without defining
	  collections
	- raw SQL can be used in queries ('get', 'find', 'count')
        Bug fixes
	- drop views everywhere that we drop tables, so views will not get in
	  way when we create tables. (arises because DROP TABLE only drops 
	  tables, not views)
	Many small cleanups in code and tests

1.19_02 2010-09-17
 	Bug fixes
	- added 'numify' to overloads in Oid and OidDeleted. arises because
	  Perls >= 5.10 use stringify to autogenerate numify
	  (Thanks to CPAN testers for finding this!)
	- found a few more places where I drop tables and now have to also
	  drop views
	Changes to tests
	- added code to emit MySQL version in 000.code_exists test
	- check MySQL version in 150.dropview test and skip if version too
	  old to support views
	- added tests for overloaded numify

1.11	2010-03-08
	Changes to build and test process only.
	Changes to Build.PL:
	- added DBD, DBI::mysql, plus some others to configure_requires. thanks 
	  to CPAN testers for finding this problem
	Changes in tests:
	- more screwing around with MySQL's max_allowed_packet variable. this
	  limits the size of the biggest object that can be stored. I have come
	  to learn that setting this variable has no effect, even though the
	  MySQL documentation claims that it is a session variable. recent
	  versions of the docs correct this misconception...

1.10	2010-03-01
	Brings CPAN version up-to-date with version used internally at ISB
        New features:
        - 'put' method added to AutoDB (in Class::AutoDB::Database) to store
           specific objects. this is a step towards cleaning up namespace of
           user-objects
        - 'oid' method added to AutoDB. this is a step towards cleaning up 

Build.PL
use strict;
no warnings;
use Cwd;
use Module::Build;

use lib qw(t);
use t::Build;			# my Module::Build subclass
my $class='t::Build';
my $MYSQL_dir=File::Spec->catdir(cwd(),qw(t MYSQL));
my $SDBM_dir=File::Spec->catdir(cwd(),qw(t SDBM));

my $builder = $class->new
  (module_name         => 'Class::AutoDB',
   license             => 'perl',
   dist_author         => q{Nat Goodman <natg@shore.net>},
   configure_requires=>{'Cwd' => 3.40,
			'Module::Build' => 0.4007,
			},
   build_requires =>   {'Carp' => 1.20,
			'Class::AutoClass' => 1.56,
			'Cwd' => 3.40,
			'DBD::mysql' => 4.024,
			'DBI' => 1.628,
			'Data::Rmap' => 0.62,
			'Fcntl' => 1.11,
			'File::Basename' => 2.82,
			'File::Path' => 2.09,
			'File::Spec' => 3.40,
			'File::Spec::Functions' => 3.40,
			'FindBin' => 1.50,
			'FreezeThaw' => 0.5001,
			'Getopt::Long' => 2.38,
			'Graph' => 0.94,
			'Hash::AutoHash::Args' => 1.18,
			'List::MoreUtils' => 0.33,
			'List::Util' => 1.23,
			'Module::Build' => 0.4007,
			'SDBM_File' => 1.09,
			'Scalar::Util' => 1.23,
			'Storable' => 2.3,
			'TAP::Harness' => 3.28,
			'TAP::Parser::Aggregator' => 3.28,
			'Test::CheckManifest' => 1.26,
			'Test::Deep' => 0.11,
			'Test::More' => 0.98,
			'Test::Pod' => 1.48,
			'Test::Pod::Content' => '0.0.6',
		       },
   requires =>         {'Class::AutoClass' => 1.56,
			'Class::Singleton' => 1.4,
			'Cwd' => 3.40,
			'DBD::mysql' => '>= 4.024, != 4.025',
			'DBI' => 1.628,
			'Exporter' => 5.68,
			'Fcntl' => 1.11,
			'Hash::AutoHash' => 1.17,
			'Hash::AutoHash::Args' => 1.18,
			'List::MoreUtils' => 0.33,
			'List::Util' => 1.32,
			'Storable' => 2.30,
			'Text::Abbrev' => 1.02,
			'Tie::ToObject' => 0.03,
			'XSLoader' => 0.15,
		       },
   recommends =>       {'Test::mysqld' => 0.17,
		       },
   add_to_cleanup      => [ 'Class-AutoDB-*',
			    File::Spec->catfile($SDBM_dir,'*'),
			    File::Spec->catfile($MYSQL_dir,'*'),
			    '*.dir', '*.pag' ],
   create_makefile_pl => 'small',
   extra_compiler_flags => ['-DUSE_PPPORT_H'],
   use_tap_harness => 1,			  
   test_files => 't/autodb.*.t',
);

$builder->create_build_script();

